packages/foam-core/src/utils/log.ts

```diff
@@ -0,0 +1,89 @@
+export interface ILogger {
+  debug(message?: any, ...params: any[]): void;
+  info(message?: any, ...params: any[]): void;
+  warn(message?: any, ...params: any[]): void;
+  error(message?: any, ...params: any[]): void;
+  getLevel(): LogLevelThreshold;
+  setLevel(level: LogLevelThreshold): void;
+}
+
+export type LogLevel = 'debug' | 'info' | 'warn' | 'error';
+export type LogLevelThreshold = LogLevel | 'off';
+
+export abstract class BaseLogger implements ILogger {
+  private static severity = {
+    debug: 1,
+    info: 2,
+    warn: 3,
+    error: 4,
+  };
+
+  constructor(private level: LogLevelThreshold = 'info') {}
+
+  abstract log(lvl: LogLevel, msg?: any, ...extra: any[]): void;
+
+  doLog(msgLevel: LogLevel, message?: any, ...params: any[]): void {
+    if (this.level === 'off') {
+      return;
+    }
+    if (BaseLogger.severity[msgLevel] >= BaseLogger.severity[this.level]) {
+      this.log(msgLevel, message, ...params);
+    }
+  }
+
+  debug(message?: any, ...params: any[]): void {
+    this.doLog('debug', message, ...params);
+  }
+  info(message?: any, ...params: any[]): void {
+    this.doLog('info', message, ...params);
+  }
+  warn(message?: any, ...params: any[]): void {
+    this.doLog('warn', message, ...params);
+  }
+  error(message?: any, ...params: any[]): void {
+    this.doLog('error', message, ...params);
+  }
+  getLevel(): LogLevelThreshold {
+    return this.level;
+  }
+  setLevel(level: LogLevelThreshold): void {
+    this.level = level;
+  }
+}
+
+export class ConsoleLogger extends BaseLogger {
```

Comment:
This and the remaining classes could be isolated into their own file

Reply:
They could be, but personally I don't mind them being in the same file. All have the same responsibility and relation, so the colocation is nice here 👍🏻

Reply:
The motivation comes from file system navigation, it becomes easier than grepping the whole project

Reply:
I think it's also a matter of convention. In Java-land what @jmg-duarte says is more the norm, in JS-land the norm is what @ingalless says.

What I like about JS is that you can have `utils/log.ts`, and as the file becomes too big turn it into a module `utils/log/index.ts`, which then allows to have more files in the directory